Translate `_d_newThrowable` hook to template

teodutu opened the pull request:
This PR lowers expressions such as `throw new E(args)` to:
```d
throw _d_newThrowable!E(args);
``` 
when compiling with `-dip1008`.

The DRuntime PRs which implement the template `_d_newThrowable` are:
- https://github.com/dlang/druntime/pull/3661
- https://github.com/dlang/druntime/pull/3692

teodutu committed:
Translate `_d_newThrowable` to template

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu committed:
dinterpret.d: Forbid evaluating `_d_newThrowable` at CTFE

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu committed:
e2ir.d: Only evaluate the false branch in `__ctfe ? ...`

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu committed:
Lower `throw new E()` to improved hook

This simplifies the lowering of `throw new E(args)` to
`throw _d_newThrowable!E(args)`.
During ctfe, the hook is interpreted like `new E(args)`.

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

thewilsonator commented:
dependant PR merged, please rebase this.

thewilsonator reviewed (commented):
is this able to be tested in this PR?

teodutu commented:
@thewilsonator Yes. This lowering is enabled by `-dip1008`. The following tests use this parameter:
- test/compilable/test20318.d
- test/fail_compilation/test20023.d
- test/runnable/test19317.d
- test/compilable/testdip1008.d

thewilsonator commented:
There are a few build kite failures, could you take a look. cc @atilaneves this fails on your projects

atilaneves commented:
> There are a few build kite failures, could you take a look. cc @atilaneves this fails on your projects

Yes - but I don't know why it'd fall on me to go and see why. I downloaded the branch from the PR anyway and tried it on unit-threaded, and it seems this PR ends up changing the purity of `std.format.spec.FormatSpec!char.FormatSpec.toString!(Appender!string).toString` to impure.

The buildkite test projects are in CI exactly to try and catch this type of breakage.

I'm more than happy to help and answer questions because of familiarity (or lack thereof) with the codebase.

RazvanN7 commented:
Rebase this and let's get it in.

teodutu committed:
testdip1008.d: Verify `@safe pure @nogc` qualifiers

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

RazvanN7 commented on src/dmd/dinterpret.d:
```diff
@@ -4866,6 +4866,66 @@ public:
+                    }
+                }
+
+                // Hack: store a ClassDeclaration instead of a StructDeclaration.
```
This seems to be storing a StructDeclaration instead of a ClassDeclaration, not the other way around.

teodutu replied:
You're right. Fixed.

RazvanN7 commented on src/dmd/dinterpret.d:
```diff
@@ -4866,6 +4866,66 @@ public:
+                auto se = ctfeEmplaceExp!StructLiteralExp(e.loc, cast(StructDeclaration) cd, elems, e.type);
+                se.origin = se;
+                se.ownedByCtfe = OwnedBy.ctfe;
+                Expression eref = ctfeEmplaceExp!ClassReferenceExp(e.loc, se, e.type);
```
Do you mind explaining why is this necessary?

teodutu replied:
`eref` needs to be a reference to `this` of the constructed class, in order to call `interpretFunction` below. This is why it needs to be a `ClassReferenceExp`. The ctor of `ClassReferenceExp` receives a `StructLiteralExp`. Hence, the need for the "hack".

MoonlightSentinel commented on src/dmd/dinterpret.d:
```diff
@@ -4866,6 +4866,66 @@ public:
+                    fieldsSoFar -= c.fields.dim;
+                    foreach (i, v; c.fields)
+                    {
+                        if (v.inuse)
+                        {
+                            e.error("circular reference to `%s`", v.toPrettyChars());
+                            result = CTFEExp.cantexp;
+                            return;
+                        }
+                        Expression m;
+                        if (v._init)
+                        {
+                            if (v._init.isVoidInitializer())
+                                m = voidInitLiteral(v.type, v).copy();
+                            else
+                                m = v.getConstInitializer(true);
+                        }
+                        else
+                            m = v.type.defaultInitLiteral(e.loc);
+                        if (exceptionOrCant(m))
+                            return;
+                        (*elems)[fieldsSoFar + i] = copyLiteral(m).copy();
+                    }
+                }
+
+                // Hack: store a ClassDeclaration instead of a StructDeclaration.
+                auto se = ctfeEmplaceExp!StructLiteralExp(e.loc, cast(StructDeclaration) cd, elems, e.type);
+                se.origin = se;
+                se.ownedByCtfe = OwnedBy.ctfe;
+                Expression eref = ctfeEmplaceExp!ClassReferenceExp(e.loc, se, e.type);
+
+                UnionExp ue = void;
+                auto err = interpretFunction(&ue, cast(FuncDeclaration) cd.ctor, istate, e.arguments, eref);
+                if (exceptionOrCant(err))
+                    return;
+
+                eref.loc = e.loc;
+                result = eref;
+                return;
+            }
```
This block seems to be copied from the class branch of `NewExp` - refactor to that code into a function instead of duplicating it here.

teodutu replied:
Wouldn't it be shorter and more expressive, albeit a bit less efficient, if I just create a `NewExp` and interpret it directly?

MoonlightSentinel commented on src/dmd/e2ir.d:
```diff
@@ -1096,8 +1096,11 @@ extern (C++) class ToElemVisitor : Visitor
             else
             {
+                assert(!(global.params.ehnogc && ne.thrownew),
+                    "This should have been rewritten to `_d_newThrowable` in the semantic phase.");
```
This assert will trigger for COM / C++ classes which are not rewritten in `expressionsem.d`?

teodutu replied:
You're right. I changed this `assert` to the same condition as in expressionsem.d: https://github.com/dlang/dmd/blob/bf6d206751c4a5bcf20ab4c19e4a7ff283a318dc/src/dmd/expressionsem.d#L3727-L3728.

RazvanN7 replied:
It doesn't seem that the `!isCOMclass() && !cd.isCPPclass()` condition did not make it.

RazvanN7 commented:
It seems that this is still failing a couple of projects on buildkite

kinke commented:
Looks like the undefined symbol errors are caused by the same underlying issue as using `-checkaction=context`. E.g., failing `unit-threaded` uses `-preview=dip1008`: https://github.com/atilaneves/unit-threaded/blob/e9e612b5126b3c99c8239f2e09b65c9303e53d94/dub.json#L51

The compiler assumes druntime and Phobos are compiled with the same switch, and so most likely skips codegen for some templates encountered in druntime/Phobos imports. So once again a template culling problem when it comes to linking libraries compiled with different switches.

RazvanN7 commented:
@kinke : @teodutu compiled druntime locally with dip1008 enabled and then compiled unit-threaded with this version however the linker error persisted. One thing we could try is to compile both druntime and phobos with dip1008 and see if that works, however, phobos uses exceptions a lot and underlying dip1008 issues might surface (I remember looking at some dip1008 issue a few years ago and my thought back then is that it works for simple cases, but there are more complex uses where it's gonna fail; IIRC it was something regarding scope exceptions that could/should have escaped their scope). But if that doesn't work, what else can we do?

kinke commented:
I don't know all the DIP1008 details, but to me it seemed to be just a way to use exceptions with `-betterC`, without GC and without druntime/Phobos. So compiling druntime/Phobos with DIP1008 doesn't seem to make much sense to me (not linked with `-betterC` anyway).

Besides fixing `-allinst`, another option might be to use `pragma(inline, true)` for `_d_newThrowable`. Not sure how that is implemented by DMD, but with LDC, that should work in the sense of truly inlining it for each all call, in each compiled object file.

teodutu commented:
@kinke Using `pragma(inline, true)` didn't solve the linking error. I guess the symbol that causes it stays the same regardless of inlining.

kinke commented:
Oh sorry, completely forgot that the problem is the nested template, `forward`. Hmm, that's called from user code too, so force-inlining that for debug builds would probably be too much. So I don't really see an alternative to `-allinst`, or some lighter, ugly special-casing of `_d_newThrowable` templates during `needsCodegen()`, similar to #13599.

teodutu committed:
Lower `throw new Error()` to new runtime hook

The new `_d_newThrowable` hook no longer calls `Error`'s ctor. Thus, the
compiler must do this instead.

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

thewilsonator commented:
Please rebase.

RazvanN7 reviewed (approved):
Don't forget to delete the old hook in druntime!

nordlow commented:
How does this improve things for the developer?

teodutu commented:
@nordlow It's pretty transparent from the perspective of the developer. The PR makes the code in the compiler easier to maintain and contributes to https://github.com/dlang/projects/issues/25. The issue only mentions arrays, but there are other cases where `TypeInfo` is used, like this one.
